### PR TITLE
[Program Plugin] Fix program UID and other bugs

### PR DESCRIPTION
**Pull request by VictoriousRaptor**
## Changes

1. Currently uid(path in this case) of `ProgramSource` is case sensitve so this can happen, which will cause duplicate results:
图片
Fixed.
2. Fixed an issue that Flow prevents UWP apps from being completely uninstalled. See https://github.com/microsoft/PowerToys/pull/5109.
3. Refactored class `ProgramSource` and removed class `DisabledProgramSource`.
4. Some small optimizations.
5. Merged #1486 because it requires changes in this pr.
6. #1532 

## Tests
- [x] Changes in #1486 and new editing dialogue logic.
- [x] No longer blocks uninstallation of uwp programs.
- [x] Duplicate paths of ProgramSource not allowed. (for both editing and adding)
- [x] UID of win32 programs and custom program sources is case-insensitve (to lowercase when `set`).
- [x] UID of Program Sources in config file can be converted to case-insensitive.
- [x] Improved index speed for custom sources (on my pc).
- [x] Disabling programs in context menu triggers a reindex.

### CODE REVIEW

File: Plugins/Flow.Launcher.Plugin.Program/Views/Models/ProgramSource.cs [hunk 1]
```diff
@@ -1,5 +1,91 @@
...
+        /// <param name="location">location of program source</param>
+        /// <param name="enabled">enabled</param>
+        public ProgramSource(string location, bool enabled=true)
+        {
+            Location = location;
+            Enabled = enabled;
+            UniqueIdentifier = location;
+        }
+
+        public ProgramSource(ProgramSource source)
```
Comment (taooceros) [inline Plugins/Flow.Launcher.Plugin.Program/Views/Models/ProgramSource.cs:43]: is it referencing the same class? Why do we need such a constructor to copy🤔?

Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Program/Views/Models/ProgramSource.cs:43]: Yes. We need it when copying instances.

File: Plugins/Flow.Launcher.Plugin.Program/Programs/UWP.cs [hunk 2]
```diff
@@ -53,21 +52,20 @@ public UWP(Package package)
...
-            AppxPackageHelper _helper = new AppxPackageHelper();
             var path = Path.Combine(Location, "AppxManifest.xml");
 
             var namespaces = XmlNamespaces(path);
             InitPackageVersion(namespaces);
 
             const uint noAttribute = 0x80;
-            const Stgm exclusiveRead = Stgm.Read | Stgm.ShareExclusive;
-            var hResult = SHCreateStreamOnFileEx(path, exclusiveRead, noAttribute, false, null, out IStream stream);
+            const Stgm nonExclusiveRead = Stgm.Read | Stgm.ShareDenyNone;
```
Comment (taooceros) [inline Plugins/Flow.Launcher.Plugin.Program/Programs/UWP.cs:61]: I see the pr you referenced completely remove the share option. Is that the same as `Stgm.ShareDenyNone`?

File: Plugins/Flow.Launcher.Plugin.Program/Main.cs [hunk 3]
```diff
@@ -203,16 +198,7 @@ private void DisableProgram(IProgram programToDelete)
                 _win32s.FirstOrDefault(x => x.UniqueIdentifier == programToDelete.UniqueIdentifier)
                     .Enabled = false;
```
Comment (taooceros) [inline Plugins/Flow.Launcher.Plugin.Program/Main.cs:199]: please append a `?` here for nullreference since we are using `FirstOrDefault`. Though, shall we use `First` instead?🤔

File: Plugins/Flow.Launcher.Plugin.Program/Views/Models/ProgramSource.cs [hunk 4]
```diff
@@ -1,5 +1,91 @@
...
+    /// <para>Custom user added program sources set UniqueIdentifier using their location</para>
+    /// </remarks>
+    public class ProgramSource
+    {
+        private string name;
+
+        public string Location { get; set; }
+        public string Name { get => name ?? new DirectoryInfo(Location).Name; set => name = value; }
+        public bool Enabled { get; set; } = true;
+        private string uid { get; set; }
```
Comment (taooceros) [inline Plugins/Flow.Launcher.Plugin.Program/Views/Models/ProgramSource.cs:22]: please make this a field.

File: Plugins/Flow.Launcher.Plugin.Program/Views/Models/ProgramSource.cs [hunk 5]
```diff
@@ -1,5 +1,91 @@
...
+            return program != null && program.UniqueIdentifier == this.UniqueIdentifier;
+        }
+
+        public override int GetHashCode()
+        {
+            return HashCode.Combine(uid);
+        }
+    }
+
+    public class DisabledProgramSource : ProgramSource
```
Comment (taooceros) [inline Plugins/Flow.Launcher.Plugin.Program/Views/Models/ProgramSource.cs:75]: I don't understand why we need an inheritance here....It only changes the mutable Property.....Could you explain your idea here?

Further, the property `Enabled` is public, so this subclass doesn't restrict people to change `Enabled` to `true`.

File: Plugins/Flow.Launcher.Plugin.Program/Views/ProgramSetting.xaml.cs [hunk 6]
```diff
@@ -199,13 +199,9 @@ private void programSourceView_Drop(object sender, DragEventArgs e)
             {
                 foreach (string directory in directories)
                 {
-                    if (Directory.Exists(directory) && !ProgramSettingDisplayList.Any(x => x.UniqueIdentifier == directory))
+                    if (Directory.Exists(directory) && !ProgramSettingDisplayList.Any(x => x.UniqueIdentifier == directory.ToLowerInvariant()))
```
Comment (taooceros) [inline Plugins/Flow.Launcher.Plugin.Program/Views/ProgramSetting.xaml.cs:202]: Please use `string.Equal` with the option Ignore case instead.

File: Plugins/Flow.Launcher.Plugin.Program/Views/Commands/ProgramSettingDisplay.cs [hunk 7]
```diff
@@ -9,20 +9,11 @@ namespace Flow.Launcher.Plugin.Program.Views.Commands
 {
     internal static class ProgramSettingDisplay
     {
-        internal static List<ProgramSource> LoadProgramSources(this List<Settings.ProgramSource> programSources)
+        internal static List<ProgramSource> LoadProgramSources(this List<ProgramSource> programSources)
```
Comment (taooceros) [inline Plugins/Flow.Launcher.Plugin.Program/Views/Commands/ProgramSettingDisplay.cs:12]: I don't think we need to copy the programsource🤔. I would suggest whether it is possible to do a bit more refactor by using `UnionBy`, but it's ok to leave it like this.

**Review by taooceros — COMMENTED** [review]
Overall it is nice. Thank you so much.

File: Plugins/Flow.Launcher.Plugin.Program/Views/Models/ProgramSource.cs (hunk 5 above)
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Program/Views/Models/ProgramSource.cs:75]: I don't quite understand either. It's not my design though, I just keep it. Maybe the original purpose was not to mess up the disabled and the enabled.

File: Plugins/Flow.Launcher.Plugin.Program/Programs/UWP.cs (hunk 2 above)
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Program/Programs/UWP.cs:61]: Yes, it's the default value.

File: Plugins/Flow.Launcher.Plugin.Program/Main.cs (hunk 3 above)
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Program/Main.cs:199]: Use `First` instead. We've checked it exists.

File: Plugins/Flow.Launcher.Plugin.Program/Views/Commands/ProgramSettingDisplay.cs (hunk 7 above)
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Program/Views/Commands/ProgramSettingDisplay.cs:12]: > I don't think we need to copy the programsource

Yep, seems like we don't need to do a deep copy.

File: Plugins/Flow.Launcher.Plugin.Program/Views/Models/ProgramSource.cs (hunk 5 above)
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Program/Views/Models/ProgramSource.cs:75]: After some research I think `DisabledProgramSource` is only for a disabled SINGLE program. If you disable a single program from context menu, a new instance is added to the list.

File: Plugins/Flow.Launcher.Plugin.Program/Programs/UWP.cs [hunk 8]
```diff
@@ -696,9 +686,14 @@ public ImageSource Logo()
...
             {
+                // TODO: Consider using infrastructure.image.imageloader?
                 if (File.Exists(path))
                 {
-                    var image = new BitmapImage(new Uri(path));
+                    var image = new BitmapImage();
+                    image.BeginInit();
+                    image.UriSource =  new Uri(path);
+                    image.CacheOption = BitmapCacheOption.OnLoad;
+                    image.EndInit();
```
Comment (taooceros) [inline Plugins/Flow.Launcher.Plugin.Program/Programs/UWP.cs:696]: please freeze it.

**Review by taooceros — COMMENTED** [review]
Overall thank you so much for doing the refactor! I will give everything a test recently (till now I just review the code). This part of code is really confusing and definitely should be refactor.

File: Plugins/Flow.Launcher.Plugin.Program/Views/Models/ProgramSource.cs (hunk 1 above)
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Program/Views/Models/ProgramSource.cs:48]: So originally it copies a instance when loading program sources from settings and copying them to the display list. It's been removed now. The only usage now is copying disabled sources to the displaylist (another problem though).

File: Plugins/Flow.Launcher.Plugin.Program/Views/Commands/ProgramSettingDisplay.cs [hunk 9]
```diff
@@ -9,75 +9,30 @@ namespace Flow.Launcher.Plugin.Program.Views.Commands
...
-                                                         Enabled = t1.Enabled
-                                                     }
-                                              ));
+                .ForEach(t1 => ProgramSetting.ProgramSettingDisplayList.Add(new ProgramSource(t1)));
         }
 
-        internal static void SetProgramSourcesStatus(this List<ProgramSource> list, List<ProgramSource> selectedProgramSourcesToDisable, bool status)
+        internal static void SetProgramSourcesStatus(List<ProgramSource> selectedProgramSourcesToDisable, bool status)
         {
             ProgramSetting.ProgramSettingDisplayList
```
[the remark this answers is not in the document]
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Program/Views/Commands/ProgramSettingDisplay.cs:37]: > we don't need to create a new list.

Can you explain it further?

File: Plugins/Flow.Launcher.Plugin.Program/Views/Models/ProgramSource.cs (hunk 5 above)
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Program/Views/Models/ProgramSource.cs]: Already removed `DisabledProgramSource`. Using `ProgramSource` instead.

File: Plugins/Flow.Launcher.Plugin.Program/Programs/Win32.cs [hunk 10]
```diff
@@ -381,6 +381,7 @@ private static string Extension(string path)
         private static IEnumerable<Win32> UnregisteredPrograms(List<ProgramSource> sources, string[] suffixes, string[] protocols)
         {
             var paths = ExceptDisabledSource(sources.Where(s => Directory.Exists(s.Location) && s.Enabled)
```
Comment (taooceros) [inline Plugins/Flow.Launcher.Plugin.Program/Programs/Win32.cs:374]: Have you tested whether parallel give performance boost? In my very early test it is not so significant somehow

File: Plugins/Flow.Launcher.Plugin.Program/FileChangeWatcher.cs [hunk 11]
```diff
@@ -1,58 +0,0 @@
-﻿using System.Collections.Generic;
-using System.IO;
-using System.Threading.Tasks;
-using Flow.Launcher.Infrastructure.Logger;
-using Flow.Launcher.Plugin.Program.Programs;
-
-namespace Flow.Launcher.Plugin.Program
-{
-    //internal static class FileChangeWatcher
```
Comment (taooceros) [inline Plugins/Flow.Launcher.Plugin.Program/FileChangeWatcher.cs:9]: why is this deleted sorry?

Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Program/FileChangeWatcher.cs:9]: It's unused. The whole class is commented long ago.

File: Plugins/Flow.Launcher.Plugin.Program/Programs/Win32.cs (hunk 10 above)
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Program/Programs/Win32.cs:374]: On my computer it is about 30% faster. I will do some more tests and show you the results.

Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Program/Programs/Win32.cs:374]: Parallel gives up to 30% boost on my computer. Not always a significant boost but generally faster.

**Comment by jjw24** [comment]
@VictoriousRaptor please resolve conflict.

**Comment by jjw24** [comment]
@VictoriousRaptor please update description on what you have tested with this change.

**Comment by VictoriousRaptor** [comment]
> @VictoriousRaptor please update description on what you have tested with this change.

Updated. Consider a version bump of the plugin?

**Comment by jjw24** [comment]
Check telhe release pr, I think this plugin already has a feature bump already.

**Review by jjw24 — APPROVED** [review]
💯 👍 👍
